build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -132,6 +132,44 @@ private void addMrjarSourceset(
         project.getTasks().withType(CheckForbiddenApisTask.class).named(forbiddenApisTaskName).configure(forbiddenApisTask -> {
             forbiddenApisTask.setIgnoreMissingClasses(true);
         });
+
+        return sourceSet;
+    }
+
+    private void createTestTask(Project project, SourceSet sourceSet, int javaVersion) {
+        var jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME);
+        project.getTasks().register(JavaPlugin.TEST_TASK_NAME + javaVersion, Test.class).configure(testTask -> {
+            testTask.dependsOn(jarTask);
+
+            SourceSetContainer sourceSets = GradleUtils.getJavaSourceSets(project);
+            FileCollection mainRuntime = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME + javaVersion).getOutput();
```

Comment:
Don't we need to remove _all_ the parent main source sets here. So if this was `test23` we'd need to remove `main21`, `main22` and `main23` or else we'll hit jar hell.

Reply:
Of course, you are right. I've fixed this to loop over the main sourcesets up to that point.